File: tools/covenant/quality_scorecard.py
#!/usr/bin/env python3
"""Agent Output Quality Scorecards.

Scores completed tasks against execution quality criteria and stores results in
cortana_quality_scores.
"""

from __future__ import annotations

import argparse
import datetime as dt
import json
import os
import re
import subprocess
from typing import Any

DB_NAME = "cortana"
DB_BIN = "/opt/homebrew/opt/postgresql@17/bin"

CRITERIA_POINTS = 20

ROLE_KEYWORDS: dict[str, tuple[str, ...]] = {
    "huragok": ("infra", "migration", "build", "tool", "automation", "service", "devops"),
    "researcher": ("research", "compare", "analysis", "findings", "sources"),
    "librarian": ("docs", "documentation", "readme", "runbook", "guide"),
    "oracle": ("forecast", "risk", "strategy", "decision", "model"),
    "monitor": ("monitor", "alert", "health", "watch", "scorecard", "quality"),
}


def sql_escape(text: str) -> str:
    return (text or "").replace("'", "''")


def run_psql(sql: str) -> str:
    env = os.environ.copy()
    env["PATH"] = f"{DB_BIN}:{env.get('PATH', '')}"
    cmd = ["psql", DB_NAME, "-q", "-X", "-v", "ON_ERROR_STOP=1", "-t", "-A", "-F", "|", "-c", sql]
    out = subprocess.run(cmd, text=True, capture_output=True, env=env)
    if out.returncode != 0:
        raise RuntimeError(out.stderr.strip() or out.stdout.strip() or "psql failed")
    return out.stdout.strip()


def run_git(args: list[str]) -> str:
    out = subprocess.run(["git", *args], text=True, capture_output=True)
    if out.returncode != 0:
        raise RuntimeError(out.stderr.strip() or out.stdout.strip() or f"git {' '.join(args)} failed")
    return out.stdout.strip()


def parse_row(raw: str) -> list[str]:
    line = next((ln for ln in (raw or "").splitlines() if ln.strip()), "")
    return line.split("|") if line else []


def infer_agent_role(*texts: str) -> str:
    blob = " ".join(t for t in texts if t).lower()
    if not blob:
        return "unknown"
    scores: dict[str, int] = {}
    for role, keywords in ROLE_KEYWORDS.items():
        hit = sum(1 for kw in keywords if kw in blob)
        if hit:
            scores[role] = hit
    if not scores:
        return "unknown"
    return sorted(scores.items(), key=lambda kv: kv[1], reverse=True)[0][0]


def get_task(task_id: int) -> dict[str, Any]:
    sql = f"""
SELECT id, status, title, description, execution_plan, COALESCE(outcome,''), COALESCE(assigned_to,''), COALESCE(metadata::text,'{{}}')
FROM cortana_tasks
WHERE id = {int(task_id)};
"""
    row = parse_row(run_psql(sql))
    if not row:
        raise ValueError(f"Task {task_id} not found")

    metadata: dict[str, Any] = {}
    try:
        metadata = json.loads(row[7]) if len(row) > 7 and row[7] else {}
    except Exception:
        metadata = {}

    return {
        "id": int(row[0]),
        "status": row[1],
        "title": row[2],
        "description": row[3],
        "execution_plan": row[4],
        "outcome": row[5],
        "assigned_to": row[6],
        "metadata": metadata,
    }


def _title_tokens(title: str) -> list[str]:
    tokens = re.findall(r"[a-z0-9]+", (title or "").lower())
    return [t for t in tokens if len(t) >= 5]


def _related_commit_hashes(task_id: int, title: str, max_commits: int = 300) -> list[str]:
    log = run_git(["log", f"-n{max_commits}", "--pretty=format:%H|%s|%b"])
    hashes: list[str] = []
    title_words = _title_tokens(title)

    for line in log.splitlines():
        parts = line.split("|", 2)
        if len(parts) < 2:
            continue
        commit_hash = parts[0]
        msg = " ".join(parts[1:]).lower()

        explicit = any(
            token in msg
            for token in (
                f"task {task_id}",
                f"task#{task_id}",
                f"#{task_id}",
                f"id {task_id}",
                f"id:{task_id}",
            )
        )
        title_hits = sum(1 for token in title_words if token in msg)

        if explicit or title_hits >= 2:
            hashes.append(commit_hash)

    return hashes


def _changed_files_for_commits(commits: list[str]) -> list[str]:
    if not commits:
        return []
    files: list[str] = []
    for commit in commits:
        out = run_git(["show", "--name-only", "--pretty=format:", commit])
        files.extend([ln.strip() for ln in out.splitlines() if ln.strip()])
    return sorted(set(files))


def _compile_python_files(files: list[str]) -> tuple[bool, list[str], str]:
    py_files = [f for f in files if f.endswith(".py") and os.path.exists(f)]
    if not py_files:
        return True, [], "No Python files changed in related commits."

    failed: list[str] = []
    last_error = ""
    for pyf in py_files:
        out = subprocess.run(["python3", "-m", "py_compile", pyf], text=True, capture_output=True)
        if out.returncode != 0:
            failed.append(pyf)
            last_error = out.stderr.strip() or out.stdout.strip()
    if failed:
        return False, failed, last_error
    return True, py_files, "All changed Python files compile."


def score(task_id: int) -> dict[str, Any]:
    task = get_task(task_id)
    commits = _related_commit_hashes(task_id, task["title"])
    changed_files = _changed_files_for_commits(commits)

    docs_required = bool(re.search(r"\b(doc|docs|documentation|readme|runbook|guide)\b", task["execution_plan"].lower()))
    docs_present = any(path.startswith("docs/") or path.lower().endswith(".md") for path in changed_files)

    compile_ok, compile_targets_or_failed, compile_details = _compile_python_files(changed_files)

    criteria = {
        "task_marked_done": {
            "passed": task["status"] == "done",
            "points": CRITERIA_POINTS if task["status"] == "done" else 0,
            "details": f"status={task['status']}",
        },
        "git_commit_made": {
            "passed": len(commits) > 0,
            "points": CRITERIA_POINTS if commits else 0,
            "details": f"commits={commits[:5]}",
        },
        "docs_created_if_required": {
            "passed": (not docs_required) or docs_present,
            "points": CRITERIA_POINTS if ((not docs_required) or docs_present) else 0,
            "details": f"docs_required={docs_required}, docs_present={docs_present}",
        },
        "python_compile_check": {
            "passed": compile_ok,
            "points": CRITERIA_POINTS if compile_ok else 0,
            "details": compile_details,
            "targets": compile_targets_or_failed,
        },
        "outcome_populated": {
            "passed": bool(task["outcome"].strip()),
            "points": CRITERIA_POINTS if task["outcome"].strip() else 0,
            "details": "outcome present" if task["outcome"].strip() else "outcome empty",
        },
    }

    total = sum(item["points"] for item in criteria.values())
    agent_role = task["assigned_to"] or infer_agent_role(task["title"], task["description"], task["execution_plan"], task["outcome"])

    insert_sql = f"""
INSERT INTO cortana_quality_scores (task_id, agent_role, score, criteria_results, scored_at)
VALUES (
  {task_id},
  '{sql_escape(agent_role)}',
  {total},
  '{sql_escape(json.dumps(criteria))}'::jsonb,
  NOW()
)
RETURNING id, scored_at;
"""
    stored = parse_row(run_psql(insert_sql))

    return {
        "task_id": task_id,
        "agent_role": agent_role,
        "score": total,
        "criteria_results": criteria,
        "commits_considered": commits,
        "changed_files": changed_files,
        "record_id": int(stored[0]) if stored else None,
        "scored_at": stored[1] if len(stored) > 1 else None,
    }


def _period_to_interval(period: str) -> str:
    if re.fullmatch(r"\d+[smhdw]", period.strip()):
        count = int(period[:-1])
        unit_map = {"s": "seconds", "m": "minutes", "h": "hours", "d": "days", "w": "weeks"}
        return f"{count} {unit_map[period[-1]]}"
    raise ValueError("Invalid period format. Use like 7d, 24h, 30m.")


def report(period: str = "7d") -> dict[str, Any]:
    interval = _period_to_interval(period)
    sql = f"""
SELECT
  agent_role,
  COUNT(*) AS samples,
  ROUND(AVG(score)::numeric, 2) AS avg_score,
  MIN(score) AS min_score,
  MAX(score) AS max_score
FROM cortana_quality_scores
WHERE scored_at >= NOW() - INTERVAL '{sql_escape(interval)}'
GROUP BY agent_role
ORDER BY avg_score DESC, samples DESC;
"""
    raw = run_psql(sql)
    rows = []
    for line in raw.splitlines():
        if not line.strip():
            continue
        role, samples, avg_score, min_score, max_score = line.split("|")
        rows.append(
            {
                "agent_role": role,
                "samples": int(samples),
                "avg_score": float(avg_score),
                "min_score": int(min_score),
                "max_score": int(max_score),
            }
        )
    return {"period": period, "results": rows}


def trends(period: str = "7d") -> dict[str, Any]:
    interval = _period_to_interval(period)
    sql = f"""
WITH recent AS (
  SELECT agent_role, AVG(score) AS avg_score, COUNT(*) AS samples
  FROM cortana_quality_scores
  WHERE scored_at >= NOW() - INTERVAL '{sql_escape(interval)}'
  GROUP BY agent_role
),
previous AS (
  SELECT agent_role, AVG(score) AS avg_score, COUNT(*) AS samples
  FROM cortana_quality_scores
  WHERE scored_at >= NOW() - (INTERVAL '{sql_escape(interval)}' * 2)
    AND scored_at < NOW() - INTERVAL '{sql_escape(interval)}'
  GROUP BY agent_role
)
SELECT
  COALESCE(r.agent_role, p.agent_role) AS agent_role,
  COALESCE(r.avg_score, 0) AS recent_avg,
  COALESCE(p.avg_score, 0) AS previous_avg,
  COALESCE(r.samples, 0) AS recent_samples,
  COALESCE(p.samples, 0) AS previous_samples,
  (COALESCE(r.avg_score, 0) - COALESCE(p.avg_score, 0)) AS delta
FROM recent r
FULL OUTER JOIN previous p ON p.agent_role = r.agent_role
ORDER BY delta DESC;
"""
    raw = run_psql(sql)
    rows = []
    for line in raw.splitlines():
        if not line.strip():
            continue
        role, recent_avg, previous_avg, recent_samples, previous_samples, delta = line.split("|")
        delta_f = float(delta)
        rows.append(
            {
                "agent_role": role,
                "recent_avg": round(float(recent_avg), 2),
                "previous_avg": round(float(previous_avg), 2),
                "recent_samples": int(recent_samples),
                "previous_samples": int(previous_samples),
                "delta": round(delta_f, 2),
                "trend": "improving" if delta_f > 0 else ("declining" if delta_f < 0 else "flat"),
            }
        )

    generated_at = dt.datetime.now(dt.timezone.utc).isoformat()
    return {"period": period, "generated_at": generated_at, "results": rows}


def cmd_score(args: argparse.Namespace) -> int:
    result = score(args.task_id)
    print(json.dumps(result, indent=2))
    return 0


def cmd_report(args: argparse.Namespace) -> int:
    result = report(period=args.period)
    print(json.dumps(result, indent=2))
    return 0


def cmd_trends(args: argparse.Namespace) -> int:
    result = trends(period=args.period)
    print(json.dumps(result, indent=2))
    return 0


def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Agent output quality scoring")
    sub = p.add_subparsers(dest="command", required=True)

    sp = sub.add_parser("score", help="Score one task and store result")
    sp.add_argument("task_id", type=int)
    sp.set_defaults(func=cmd_score)

    rp = sub.add_parser("report", help="Aggregate quality scores by role")
    rp.add_argument("--period", default="7d", help="Lookback period like 7d, 24h")
    rp.set_defaults(func=cmd_report)

    tp = sub.add_parser("trends", help="Compare recent vs previous period")
    tp.add_argument("--period", default="7d", help="Window period like 7d, 24h")
    tp.set_defaults(func=cmd_trends)

    return p


def main() -> int:
    parser = build_parser()
    args = parser.parse_args()
    return int(args.func(args))


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/covenant/quality_scorecard.ts
#!/usr/bin/env npx tsx

/** Agent Output Quality Scorecards. */

import fs from "fs";
import { spawnSync } from "child_process";
import { runPsql, withPostgresPath } from "../lib/db.js";

const CRITERIA_POINTS = 20;

const ROLE_KEYWORDS: Record<string, string[]> = {
  huragok: ["infra", "migration", "build", "tool", "automation", "service", "devops"],
  researcher: ["research", "compare", "analysis", "findings", "sources"],
  librarian: ["docs", "documentation", "readme", "runbook", "guide"],
  oracle: ["forecast", "risk", "strategy", "decision", "model"],
  monitor: ["monitor", "alert", "health", "watch", "scorecard", "quality"],
};

type Json = Record<string, any>;

function sqlEscape(text: string): string {
  return (text || "").replace(/'/g, "''");
}

function runPsqlQuery(sql: string): string {
  const result = runPsql(sql, {
    db: "cortana",
    args: ["-q", "-X", "-v", "ON_ERROR_STOP=1", "-t", "-A", "-F", "|"],
    env: withPostgresPath(process.env),
  });
  if (result.status !== 0) {
    const err = (result.stderr || result.stdout || "psql failed").toString().trim();
    throw new Error(err || "psql failed");
  }
  return (result.stdout || "").toString().trim();
}

function runGit(args: string[]): string {
  const out = spawnSync("git", args, { encoding: "utf8" });
  if (out.status !== 0) {
    const msg = (out.stderr || out.stdout || `git ${args.join(" ")} failed`).toString().trim();
    throw new Error(msg || `git ${args.join(" ")} failed`);
  }
  return (out.stdout || "").toString().trim();
}

function parseRow(raw: string): string[] {
  const line = (raw || "")
    .split(/\r?\n/)
    .find((ln) => ln.trim()) || "";
  return line ? line.split("|") : [];
}

function inferAgentRole(...texts: string[]): string {
  const blob = texts.filter((t) => t).join(" ").toLowerCase();
  if (!blob) return "unknown";
  const scores: Record<string, number> = {};
  for (const [role, keywords] of Object.entries(ROLE_KEYWORDS)) {
    const hit = keywords.filter((kw) => blob.includes(kw)).length;
    if (hit) scores[role] = hit;
  }
  const best = Object.entries(scores).sort((a, b) => b[1] - a[1])[0];
  return best ? best[0] : "unknown";
}

function getTask(taskId: number): Json {
  const sql = `
SELECT id, status, title, description, execution_plan, COALESCE(outcome,''), COALESCE(assigned_to,''), COALESCE(metadata::text,'{}')
FROM cortana_tasks
WHERE id = ${Math.trunc(taskId)};
`;
  const row = parseRow(runPsqlQuery(sql));
  if (!row.length) {
    throw new Error(`Task ${taskId} not found`);
  }

  let metadata: Json = {};
  try {
    metadata = row[7] ? JSON.parse(row[7]) : {};
  } catch {
    metadata = {};
  }

  return {
    id: Number(row[0]),
    status: row[1],
    title: row[2],
    description: row[3],
    execution_plan: row[4],
    outcome: row[5],
    assigned_to: row[6],
    metadata,
  };
}

function titleTokens(title: string): string[] {
  const tokens = (title || "").toLowerCase().match(/[a-z0-9]+/g) || [];
  return tokens.filter((t) => t.length >= 5);
}

function relatedCommitHashes(taskId: number, title: string, maxCommits = 300): string[] {
  const log = runGit(["log", `-n${maxCommits}`, "--pretty=format:%H|%s|%b"]);
  const hashes: string[] = [];
  const titleWords = titleTokens(title);

  for (const line of log.split(/\r?\n/)) {
    const parts = line.split("|", 3);
    if (parts.length < 2) continue;
    const commitHash = parts[0];
    const msg = [parts[1], parts[2] || ""].join(" ").toLowerCase();

    const explicit = [
      `task ${taskId}`,
      `task#${taskId}`,
      `#${taskId}`,
      `id ${taskId}`,
      `id:${taskId}`,
    ].some((token) => msg.includes(token));

    const titleHits = titleWords.filter((token) => msg.includes(token)).length;

    if (explicit || titleHits >= 2) hashes.push(commitHash);
  }

  return hashes;
}

function changedFilesForCommits(commits: string[]): string[] {
  if (!commits.length) return [];
  const files: string[] = [];
  for (const commit of commits) {
    const out = runGit(["show", "--name-only", "--pretty=format:", commit]);
    for (const line of out.split(/\r?\n/)) {
      if (line.trim()) files.push(line.trim());
    }
  }
  return Array.from(new Set(files)).sort();
}

function compilePythonFiles(files: string[]): [boolean, string[], string] {
  const pyFiles = files.filter((f) => f.endsWith(".py") && fs.existsSync(f));
  if (!pyFiles.length) {
    return [true, [], "No Python files changed in related commits."];
  }

  const failed: string[] = [];
  let lastError = "";
  for (const pyf of pyFiles) {
    const out = spawnSync("python3", ["-m", "py_compile", pyf], { encoding: "utf8" });
    if (out.status !== 0) {
      failed.push(pyf);
      lastError = (out.stderr || out.stdout || "").toString().trim();
    }
  }

  if (failed.length) return [false, failed, lastError];
  return [true, pyFiles, "All changed Python files compile."];
}

function score(taskId: number): Json {
  const task = getTask(taskId);
  const commits = relatedCommitHashes(taskId, task.title);
  const changedFiles = changedFilesForCommits(commits);

  const docsRequired = /\b(doc|docs|documentation|readme|runbook|guide)\b/i.test(
    String(task.execution_plan || "")
  );
  const docsPresent = changedFiles.some(
    (p) => p.startsWith("docs/") || p.toLowerCase().endsWith(".md")
  );

  const [compileOk, compileTargetsOrFailed, compileDetails] = compilePythonFiles(changedFiles);

  const commitsPreview = `[${commits
    .slice(0, 5)
    .map((c) => `'${c}'`)
    .join(", ")}]`;

  const criteria: Json = {
    task_marked_done: {
      passed: task.status === "completed",
      points: task.status === "completed" ? CRITERIA_POINTS : 0,
      details: `status=${task.status}`,
    },
    git_commit_made: {
      passed: commits.length > 0,
      points: commits.length ? CRITERIA_POINTS : 0,
      details: `commits=${commitsPreview}`,
    },
    docs_created_if_required: {
      passed: !docsRequired || docsPresent,
      points: !docsRequired || docsPresent ? CRITERIA_POINTS : 0,
      details: `docs_required=${docsRequired}, docs_present=${docsPresent}`,
    },
    python_compile_check: {
      passed: compileOk,
      points: compileOk ? CRITERIA_POINTS : 0,
      details: compileDetails,
      targets: compileTargetsOrFailed,
    },
    outcome_populated: {
      passed: Boolean(String(task.outcome || "").trim()),
      points: String(task.outcome || "").trim() ? CRITERIA_POINTS : 0,
      details: String(task.outcome || "").trim() ? "outcome present" : "outcome empty",
    },
  };

  const total = Object.values(criteria).reduce((sum, item) => sum + Number(item.points || 0), 0);
  const agentRole = task.assigned_to || inferAgentRole(task.title, task.description, task.execution_plan, task.outcome);

  const insertSql = `
INSERT INTO cortana_quality_scores (task_id, agent_role, score, criteria_results, scored_at)
VALUES (
  ${Math.trunc(taskId)},
  '${sqlEscape(agentRole)}',
  ${total},
  '${sqlEscape(JSON.stringify(criteria))}'::jsonb,
  NOW()
)
RETURNING id, scored_at;
`;
  const stored = parseRow(runPsqlQuery(insertSql));

  return {
    task_id: taskId,
    agent_role: agentRole,
    score: total,
    criteria_results: criteria,
    commits_considered: commits,
    changed_files: changedFiles,
    record_id: stored.length ? Number(stored[0]) : null,
    scored_at: stored.length > 1 ? stored[1] : null,
  };
}

function periodToInterval(period: string): string {
  const match = period.trim().match(/^(\d+)([smhdw])$/);
  if (!match) throw new Error("Invalid period format. Use like 7d, 24h, 30m.");
  const count = Number(match[1]);
  const unitMap: Record<string, string> = {
    s: "seconds",
    m: "minutes",
    h: "hours",
    d: "days",
    w: "weeks",
  };
  return `${count} ${unitMap[match[2]]}`;
}

function report(period = "7d"): Json {
  const interval = periodToInterval(period);
  const sql = `
SELECT
  agent_role,
  COUNT(*) AS samples,
  ROUND(AVG(score)::numeric, 2) AS avg_score,
  MIN(score) AS min_score,
  MAX(score) AS max_score
FROM cortana_quality_scores
WHERE scored_at >= NOW() - INTERVAL '${sqlEscape(interval)}'
GROUP BY agent_role
ORDER BY avg_score DESC, samples DESC;
`;
  const raw = runPsqlQuery(sql);
  const rows: Json[] = [];
  for (const line of raw.split(/\r?\n/)) {
    if (!line.trim()) continue;
    const [role, samples, avgScore, minScore, maxScore] = line.split("|");
    rows.push({
      agent_role: role,
      samples: Number(samples),
      avg_score: Number(avgScore),
      min_score: Number(minScore),
      max_score: Number(maxScore),
    });
  }
  return { period, results: rows };
}

function trends(period = "7d"): Json {
  const interval = periodToInterval(period);
  const sql = `
WITH recent AS (
  SELECT agent_role, AVG(score) AS avg_score, COUNT(*) AS samples
  FROM cortana_quality_scores
  WHERE scored_at >= NOW() - INTERVAL '${sqlEscape(interval)}'
  GROUP BY agent_role
),
previous AS (
  SELECT agent_role, AVG(score) AS avg_score, COUNT(*) AS samples
  FROM cortana_quality_scores
  WHERE scored_at >= NOW() - (INTERVAL '${sqlEscape(interval)}' * 2)
    AND scored_at < NOW() - INTERVAL '${sqlEscape(interval)}'
  GROUP BY agent_role
)
SELECT
  COALESCE(r.agent_role, p.agent_role) AS agent_role,
  COALESCE(r.avg_score, 0) AS recent_avg,
  COALESCE(p.avg_score, 0) AS previous_avg,
  COALESCE(r.samples, 0) AS recent_samples,
  COALESCE(p.samples, 0) AS previous_samples,
  (COALESCE(r.avg_score, 0) - COALESCE(p.avg_score, 0)) AS delta
FROM recent r
FULL OUTER JOIN previous p ON p.agent_role = r.agent_role
ORDER BY delta DESC;
`;
  const raw = runPsqlQuery(sql);
  const rows: Json[] = [];
  for (const line of raw.split(/\r?\n/)) {
    if (!line.trim()) continue;
    const [role, recentAvg, previousAvg, recentSamples, previousSamples, delta] = line.split("|");
    const deltaNum = Number(delta);
    rows.push({
      agent_role: role,
      recent_avg: Number(Number(recentAvg).toFixed(2)),
      previous_avg: Number(Number(previousAvg).toFixed(2)),
      recent_samples: Number(recentSamples),
      previous_samples: Number(previousSamples),
      delta: Number(deltaNum.toFixed(2)),
      trend: deltaNum > 0 ? "improving" : deltaNum < 0 ? "declining" : "flat",
    });
  }

  const generatedAt = new Date().toISOString().replace("Z", "+00:00");
  return { period, generated_at: generatedAt, results: rows };
}

function usageError(): never {
  console.error("usage: quality_scorecard.py score <task_id> | report [--period 7d] | trends [--period 7d]");
  process.exit(2);
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const cmd = args.shift();
  if (!cmd) usageError();

  if (cmd === "score") {
    const taskId = args[0];
    if (!taskId) usageError();
    console.log(JSON.stringify(score(Number(taskId)), null, 2));
    return;
  }

  const periodIdx = args.indexOf("--period");
  const period = periodIdx >= 0 ? args[periodIdx + 1] : "7d";

  if (cmd === "report") {
    console.log(JSON.stringify(report(period), null, 2));
    return;
  }
  if (cmd === "trends") {
    console.log(JSON.stringify(trends(period), null, 2));
    return;
  }

  usageError();
}

main().catch((err) => {
  console.error(err instanceof Error ? err.stack || err.message : String(err));
  process.exit(1);
});
